Using the same analysis pathway as described for HBV, the demultiplexed HiFi reads can be aligned to the human GENCODE reference genome (Human B38 GenCode fasta file, www.gencodegenes.org) with low quality reads (MAPQ scores < 30) removed.

In the manuscript we chose not to assess the effect of our xGen enrichment on the human cDNA transcriptome as we did not have a control pre-enrichment data set for comparison.

If one carries out the human analysis on can findt many human transcripts that were differentially expressed between the infection and transfection samples (R script quantify_human.R). Gene set enrichment analysis (GSEA) of these results suggested these genes were involved in various gene ontology biological processes. In infected cells we observed an over representation of genes associated with pathways pertaining to the mitotic cell cycle, and chromosome and other nucleus-associated pathways. In contrast, transfected cells showed a perturbation of genes involved in membrane regulation, perhaps reflecting the lipid delivery of a DNA bolus. However we feel that without the essential controls, these data are difficult to interpret.
